src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -144,52 +215,79 @@ sealed trait LazyModuleImpLike extends RawModule
   val auto: AutoBundle
   protected[diplomacy] val dangles: Seq[Dangle]
 
-  // .module had better not be accessed while LazyModules are still being built!
-  require (!LazyModule.scope.isDefined, s"${wrapper.name}.module was constructed before LazyModule() was run on ${LazyModule.scope.get.name}")
+  /** [[wrapper.module]] had better not be accessed while LazyModules are still being built! */
+  require (LazyModule.scope.isEmpty, s"${wrapper.name}.module was constructed before LazyModule() was run on ${LazyModule.scope.get.name}")
 
+  /** set name from [[wrapper]]. */
   override def desiredName = wrapper.desiredName
   suggestName(wrapper.suggestedName)
 
-  implicit val p = wrapper.p
+  /** [[Parameters]] for diplomatic [[Module]]s. */
+  implicit val p: Parameters = wrapper.p
 
+  /** [[instantiate]] will be called when a instance of [[LazyModuleImp]] is created. */
   protected[diplomacy] def instantiate() = {
     val childDangles = wrapper.children.reverse.flatMap { c =>
       implicit val sourceInfo = c.info
+      /** calling [[c.module]] will push the real [[Module]] into [[chisel3.internal.Builder]],
+        * notice this place is not calling [[instantiate]], it just push the [[c.module]] to Builder.
+        * */
       val mod = Module(c.module)
+      /** ask each child to finish instantiate. */
       mod.finishInstantiate()
+      /** return dangles of each child. */
       mod.dangles
     }
+
+    /** ask each node in the [[LazyModule]] to call [[BaseNode.instantiate]].
```

Comment:
this set of comments isn't super helpful yet. What is a dangle? 

I think a dangle is "all the Edge connections left over after internal connections have been made, which would be passed up to the next level of Lazy Module hierarchy for connection". Does that sound right? I've tried to edit the below to match my understanding if that is the case.

Reply:
> I think a dangle is "all the Edge connections left over after internal connections have been made, which would be passed up to the next level of Lazy Module hierarchy for connection". Does that sound right.

I don't think so, a `Dangle` will be generated in `Nodes.scala` with `danglesOut` and `danglesIn` method. Thus `Dangle` is that case are not connected yet.
Both `LazyModule.instantiate._2` and `Base.instantiate` will return `Dangle`, I think it's a wrapper to connection, which contains 
1. the port index in sink and source(`HalfEdge`)
2. name
3. Bundle(`data`)